### PR TITLE
e2e with IBC tx and test

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #1201 

## Description

Please note, a majority of this is the original work of @alexanderbez and this is a simple port over from his work on gaia to osmosis.

- Added IBC functionality through hermes and added a single IBC test to ensure the tx was received on chainB from chainA
- Changed from `name` to `id` as requested by Dev
- Commented out TestQueryBalance. This is because we are currently relying on the order of tests for this test to pass. As we develop further understanding of how we want these tests to be ran, we can re-enable this and/or use it elsewhere.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: tests/e2e/e2e_util_test.go
Comment (p0mvn) [inline tests/e2e/e2e_util_test.go:160]: Seems to be unused?

File: tests/e2e/e2e_util_test.go
Comment (p0mvn) [inline tests/e2e/e2e_util_test.go:139]: Is this used?

File: tests/e2e/e2e_util_test.go
Comment (p0mvn) [inline tests/e2e/e2e_util_test.go:114]: Not used yet? Should we start committing things only if we are actually using them in tests?

File: tests/e2e/e2e_util_test.go
Comment (p0mvn) [reply tests/e2e/e2e_util_test.go:160]: Also, from what I remember this wouldn't work. I was adapting `TestQueryBalances` from this test, and I think it wouldn't work in osmosis if we were to run it

File: tests/e2e/scripts/hermes_bootstrap.sh
Comment (p0mvn) [inline tests/e2e/scripts/hermes_bootstrap.sh:72]: new line

File: tests/e2e/e2e_test.go
Comment (p0mvn) [inline tests/e2e/e2e_test.go:14]: I'd prefer if we make this work right away instead of committing dead code that previously worked.

I think for now we should figure out how we can enforce a specific order and let tests run after the other so that we can expect specific values. Later, we can refactor in terms of setup/teardown before each test.

What do you think?

**Review by p0mvn — CHANGES_REQUESTED** [review]
Great work on this. Left some comments, please take a look

Also

File: tests/e2e/docker/hermes.Dockerfile
Comment (czarcas7ic) [reply tests/e2e/docker/hermes.Dockerfile:1]: Yes there is, great point, will update

File: tests/e2e/e2e_setup_test.go
Comment (czarcas7ic) [reply tests/e2e/e2e_setup_test.go:373]: @nikever do you know how to add this as an image to our repo? I'm not positive of the process

File: Makefile
Comment (czarcas7ic) [reply Makefile:241]: @nikever can we link on getting this on our docker hub as well?

File: tests/e2e/e2e_util_test.go
Comment (czarcas7ic) [reply tests/e2e/e2e_util_test.go:160]: agreed, removed

File: tests/e2e/e2e_util_test.go
Comment (czarcas7ic) [reply tests/e2e/e2e_util_test.go:139]: no, removed!

File: tests/e2e/e2e_util_test.go
Comment (czarcas7ic) [reply tests/e2e/e2e_util_test.go:114]: agreed, removed as well!

File: Makefile
Comment (p0mvn) [reply Makefile:241]: I would also like to learn about how to get access to our docker hub. Please let me know if it's possible to get permissions for that

File: tests/e2e/e2e_setup_test.go
Comment (p0mvn) [reply tests/e2e/e2e_setup_test.go:373]: To avoid blocking this PR, let's create an issue to address this separately? 

It should be fine to use the `cosmos` image temporarily, The main concern about depending on external images, especially with the latest tag, is that whenever they make a change, our tests might get broken

Comment (nikever) [reply tests/e2e/e2e_setup_test.go:373]: The credentials for the `osmobot` Docker account are saved as GitHub Secrets. You can use them to build the Image via CI as in https://github.com/osmosis-labs/osmosis/blob/main/.github/workflows/docker.yml#L24-L29

I will also create personal accounts for both of you on our Docker account.

Lets use a new repository `osmolabs/osmosis-e2e` for all these new images

File: tests/e2e/e2e_test.go
Comment (czarcas7ic) [reply tests/e2e/e2e_test.go:14]: Sounds good! I got it to work, but I decided to remove gas fees in order to make calculations simpler (still kept the const so we can add fees if we want in the future)

File: tests/e2e/e2e_test.go
Comment (p0mvn) [inline tests/e2e/e2e_test.go:32]: let's restore this

File: tests/e2e/e2e_test.go
Comment (p0mvn) [inline tests/e2e/e2e_test.go:26]: and this one

File: tests/e2e/e2e_test.go
Comment (p0mvn) [reply tests/e2e/e2e_test.go]: Looking good. Yeah let's add this later

**Review by p0mvn — APPROVED** [review]
Please restore commented-out lines in `TestQueryBalances`. Other than that LGTM!

File: Makefile
Comment (alexanderbez) [reply Makefile:241]: Yes, good idea. It's very easy to push to our org. All you need is a github PAT (personal access token) and then you login with that and just push it. Note, just make sure the package set to public.

File: tests/e2e/e2e_setup_test.go
Comment (alexanderbez) [reply tests/e2e/e2e_setup_test.go:403]: Validators? You mean the node containers themselves? I believe we already have a health check for that. Are you suggesting that we implement a test that queries for validators?

File: tests/e2e/e2e_test.go
Comment (alexanderbez) [reply tests/e2e/e2e_test.go:26]: So why is this commented out? Is this because or test execution order?

**Review by alexanderbez — APPROVED** [review]
🎉 🎉 🎉

File: tests/e2e/e2e_test.go
Comment (alexanderbez) [reply tests/e2e/e2e_test.go]: Yeah, I'm a bit torn. Typically, the purist in me would state that test execution order should be independent. However, there are exceptions, and I feel like integration/e2e tests where there is a single global state is certainly one of them.

The only way I know for explicitly defining test execution order is by the test name as testify's suite package executes them based on alphabetical order. We can probably achieve this via a prefix and/or suffix to each test name.

**Comment by czarcas7ic** [comment]
Had a call with @nikever, he was able to get our own hosted hermes image and integrated it into the e2e test. He will now attempt a small modification on the gh action and after he pushes this we can merge this pr

File: Makefile
Comment (p0mvn) [inline Makefile]: Should we still have this for convenience to build locally?

File: tests/e2e/docker/hermes.Dockerfile
Comment (p0mvn) [inline tests/e2e/docker/hermes.Dockerfile:1]: Why would we host our own `hermes` image?

I think we should only store `hermes-e2e` image because this is something with regards to the e2e tests that are "under our control". We also have Dockerfile for building the `hermes-e2e` image in this repository.

On the other hand, we are not developing or building hermes itself. So I think we should just reuse their image. Then, when there is a new version, we can just bump it up here. If we store our own, we have to upload a new one to our docker hub every time, and I don't really see a reason for doing that.

Let me know what you think

File: tests/e2e/e2e_setup_test.go
Comment (p0mvn) [inline tests/e2e/e2e_setup_test.go:375]: shouldn't this be `hermes-e2e`?

**Review by p0mvn — CHANGES_REQUESTED** [review]

File: tests/e2e/e2e_setup_test.go
Comment (nikever) [reply tests/e2e/e2e_setup_test.go:375]: I don't think so.

They called it `hermes-e2e` because it was a modified version of the official `hermes` to be able to run tests (mainly due to permission problems). 

WIth the changes made we can use the official  `hermes`  image to run tests saving us the time needed to build the `hermes-e2e` image.

File: tests/e2e/docker/hermes.Dockerfile
Comment (nikever) [reply tests/e2e/docker/hermes.Dockerfile]: Since we don't need to build the `hermes-e2e` image anymore we can use the official `hermes` image.

If for whatever reason in the future we need to deploy a specific `hermes-e2e` image for our e2e tests we can undo this change but I would try to stay as close to upstream as possible.

Regarding the other point. `osmolabs/hermes` it's just a mirror of the official `hermes` page. 
We could use the upstream one ofc but it's nice to have control of the artifacts that we use (e.g. they make the repo private / it is inaccessible / etc...). The burden of mantaining the mirror is minimal I think

**Comment by nikever** [comment]
e2e running correctly with the last changes: https://github.com/osmosis-labs/osmosis/runs/5889286653?check_suite_focus=true 

🚀🚀

**Review by p0mvn — APPROVED** [review]
:ship:

File: tests/e2e/e2e_util_test.go
Comment (adizere) [inline tests/e2e/e2e_util_test.go:27]: Note that we plan to change the `create channel` invocation in Hermes. From

`create channel <CHAIN_A> <CHAIN_B> --port-a=transfer --port-b=transfert`

to 

`create channel <CHAIN-A-ID> <CONNECTION-A-ID> --port-a <PORT-ID> --port-b <PORT-ID>`.

In other words, you need to (1) `create connection` and then (2) pass the connection identifier when you invoke `create channel`. This mean you would need to adapt the CI to add an additional CLI command for creating a new connection, as follows:

`create connection s.chainA.id s.chainB.id`.

Assuming the chains are newly-spawned, the connection identifier will be `connection-0` on both chains. Then you would need to invoke create channel as follows:

`create channel s.chainA.id connection-0 --port-a=transfer --port-b=transfer`

Comment (adizere) [reply tests/e2e/e2e_util_test.go:27]: PS. This change was motivated by relayer operators abusing the command (we got complaints that people create many channels in production with varying client security parameters, ref: https://github.com/informalsystems/ibc-rs/issues/1421).

Would be glad to take your requirements into account and do stuff differently if you have suggestions to change this.

Comment (ValarDragon) [reply tests/e2e/e2e_util_test.go:27]: This seems like a great change to me! Thanks for letting us know / good design decision!

Comment (adizere) [reply tests/e2e/e2e_util_test.go:27]: Great to hear it won't affect your setup adversely.

The change will land in Hermes v0.14, slated for end of April.

**Pull request merged**
